### PR TITLE
Use `MongoTemplate` to enable document update without full rewrite

### PR DESCRIPTION
**Pull request by jbern0rd**

### CODE REVIEW

File: src/main/java/com/iexec/core/task/TaskService.java [hunk 1]
```diff
@@ -126,6 +134,24 @@ public Optional<Task> updateTask(Task task) {
...
+                Task.class);
+        log.debug("Updated chainTaskId {} {}", task.getChainTaskId(), result);
+        return getTaskByChainTaskId(task.getChainTaskId());
+    }
+
+    public void updateTask(String chainTaskId, Update update) {
+        UpdateResult result = mongoTemplate.updateFirst(
+                Query.query(Criteria.where("chainTaskId").is(chainTaskId)),
+                update, Task.class);
+        log.debug("Updated chainTaskId {} {}", chainTaskId, result);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/task/TaskService.java]: please see d4d9a22

File: src/main/java/com/iexec/core/task/TaskService.java [hunk 2]
```diff
@@ -126,6 +134,24 @@ public Optional<Task> updateTask(Task task) {
...
     }
 
+    public Optional<Task> updateTaskStatus(Task task, List<TaskStatusChange> statusChanges) {
+        Update update = Update.update("currentStatus", task.getCurrentStatus());
+        update.push("dateStatusList").each(statusChanges);
+        UpdateResult result = mongoTemplate.updateFirst(
+                Query.query(Criteria.where("chainTaskId").is(task.getChainTaskId())),
+                update,
+                Task.class);
+        log.debug("Updated chainTaskId {} {}", task.getChainTaskId(), result);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/task/TaskService.java]: please see d4d9a22

File: src/main/java/com/iexec/core/task/update/TaskUpdateManager.java [hunk 3]
```diff
@@ -206,42 +208,48 @@ void updateTask(String chainTaskId) {
...
-        TaskStatus lastStatus = statuses[statuses.length - 1];
+        TaskStatus currentStatus = task.getCurrentStatus();
+        List<TaskStatusChange> statusChanges = new ArrayList<>();
         for (TaskStatus newStatus : statuses) {
             log.info("Create TaskStatusChange succeeded [chainTaskId:{}, currentStatus:{}, newStatus:{}]",
                     task.getChainTaskId(), task.getCurrentStatus(), newStatus);
-            task.changeStatus(newStatus, null);
+            final TaskStatusChange statusChange = TaskStatusChange.builder().status(newStatus).build();
+            task.setCurrentStatus(newStatus);
+            task.getDateStatusList().add(statusChange);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/task/update/TaskUpdateManager.java:218]: agreement to keep it as is for the moment as the `changeStatus` method is only used in tests

File: src/main/java/com/iexec/core/worker/WorkerService.java [hunk 4]
```diff
@@ -308,14 +318,12 @@ public Optional<Worker> removeChainTaskIdFromWorker(String chainTaskId, String w
...
-            worker.removeChainTaskId(chainTaskId);
-            log.info("Removed chainTaskId from worker [chainTaskId:{}, walletAddress:{}]", chainTaskId, walletAddress);
-            return Optional.of(workerRepository.save(worker));
-        }
-        return Optional.empty();
+        UpdateResult result = mongoTemplate.updateFirst(
+                Query.query(Criteria.where("walletAddress").is(walletAddress)),
+                new Update().pull("computingChainTaskIds", chainTaskId).pull("participatingChainTaskIds", chainTaskId),
+                Worker.class);
+        log.info("Remove chainTaskId {} {}", chainTaskId, result);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/worker/WorkerService.java]: please see 0c3b294

File: src/main/java/com/iexec/core/worker/WorkerService.java [hunk 5]
```diff
@@ -326,14 +334,12 @@ public Optional<Worker> removeComputedChainTaskIdFromWorker(String chainTaskId,
...
-            worker.removeComputedChainTaskId(chainTaskId);
-            log.info("Removed computed chainTaskId from worker [chainTaskId:{}, walletAddress:{}]", chainTaskId, walletAddress);
-            return Optional.of(workerRepository.save(worker));
-        }
-        return Optional.empty();
+        UpdateResult result = mongoTemplate.updateFirst(
+                Query.query(Criteria.where("walletAddress").is(walletAddress)),
+                new Update().pull("computingChainTaskIds", chainTaskId),
+                Worker.class);
+        log.info("Remove computed chainTaskId {} {}", chainTaskId, result);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/worker/WorkerService.java]: please see 0c3b294

**Review by mcornaton — APPROVED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 
 ## [[NEXT]](https://github.com/iExecBlockchainComputing/iexec-core/releases/tag/vNEXT) 2024
 
+### New Features
+
+- Use `MongoTemplate` to enable document update withour full rewrite. (#661)
```
Comment (mcornaton) [inline CHANGELOG.md:9]:
```suggestion
- Use `MongoTemplate` to enable document update without full rewrite. (#661)
```

**Review by thewhitewizard — APPROVED** [review]

File: CHANGELOG.md (hunk 6 above)
Comment (jbern0rd) [reply CHANGELOG.md]: see 0d2ae72

**Pull request merged**